### ISSUE TITLE
FOF30 library shows up in 'Extension ->Discover' with older version

### ISSUE THREAD

**Issue by jjnxpct** [joomla/joomla-cms#23720]
### Steps to reproduce the issue

Have the latest version of Joomla (3.9.2) and recent versions of Akeeba Backup or Admin Tools.

### Expected result
In the extension manager the FOF30 gives us version 3.3.9. I don't see why it would show up in de 'discover' section of the extension manager.

### Actual result
It show up in the 'discover' section of the extension manager. But version number is 3.3.6. So older version.

### System information (as much as possible)
Joomla 3.9.2
PHP 7.2
Latest versions of Akeeba Backup and Admin Tools pro.

### Additional comments

I have contacted Akeeba. They told me about the changes form 'library to file' and about a bug and rules that prevent them from fixing this. It feel like there have been some discussions on this issue before and it might not have been resolved to everyone's liking... I don't want to cause any (new) annoyances. I just want to check if there is a way to prevent this discovered library to show up in the discover section. Or if there is a bug in the way this 'discover' or install process works then maybe it can be addressed in version 4? If not we'll just ignore this and focus on other issues.
[https://inxpactdrive.nl/s/mFSIKprjC85yzaH](url)
[https://inxpactdrive.nl/s/RwRN4gT2YfQa64S](url)

**Comment by ghost** [comment]
@HLeithner can you please comment?

**Comment by HLeithner** [comment]
@mbabker do you know the reason for this and if its fixable?

**Comment by ghost** [comment]
@mbabker can you please comment?

**Comment by HLeithner** [comment]
@nikosdion could you give me a hint or an issue id where this was discussed?

**Comment by nikosdion** [comment]
These are two different extensions. The old one is a "library" extension (lib_fof30). The newer one is a "files" (files_fof30) extension. 

We changed from a library extension to a files extension because of the braindead way Joomla! handles library extensions on upgrade. The workflow is delete the entire lilbrary directory first(!!!), _then_ run the preinstall method of the package's script and if it doesn't return false continue with installing the new version.

The reason why this is braindead is that the preinstall method is supposed to check if the Joomla and PHP versions are acceptable (since com_installer does not have a way to check that by itself) and return false if the minimum requirements are not met. See the problem now? If you try to install a version of the library that requires a newer PHP and / or Joomla version your library is deleted (but not uninstalled). Therefore all your extensions, **including system plugins**, fail hard. When I say "hard" I mean "the site is down with a PHP fatal error".

The issue and comment I referenced in my discussion to the user is https://github.com/joomla/joomla-cms/issues/16720#issuecomment-309094471 and https://github.com/joomla/joomla-cms/issues/16720#issuecomment-309096114

The reason why both manifests are present is that you cannot cleanly switch from a library package to a files package. I try to clean up leftover files but that is far from being a robust solution. Sticking with a library package is even worse since we had users with broken sites after trying to update, an issue which would only get worse as we were getting ready to drop support for Joomla! 3.4 to 3.7 inclusive and PHP 5.4 and 5.5. Uninstalling the library package is not going to work either, since the script would kick in and prevent you from doing that due to existing dependencies. Uninstalling all of our software is not a solution either because the user would lose all their settings. 

Basically, this nuissance is the direct result of a Joomla bug making it impossible for developers to use the correct package type required and having to do a "hot swap" to the wrong kind of extension to work around it. Even if Joomla fixes the library update issue it won't solve the existing users' issue with the dual discovery of the "same" (but actually different) extension.

**Comment by HLeithner** [comment]
thx for your explanation, is there a manual way to fix it?

**Comment by nikosdion** [comment]
Without having tried it, I believe that you can delete administrator/manifests/libraries/lib_fof30.xml and then install the newest FOF 3 version. I _think_ that Discover does not look into the currently installed extensions so it won't complain again. But, really, there is no good way manual or automatic. If there was a good manual way to do it I'd have already automated it and have it run during the installation of the new files type package.

**Comment by HLeithner** [comment]
thx @nikosdion

@jjnxpct would you like to try this and I close the issue. Fixing the update should be an own issue.
